_posts/2017-11-22-macros.md

```diff
@@ -0,0 +1,305 @@
+
+> One attractive aspect of the system is that it works equally well for staging
+> and for macros.  But as a macro system it is quite restrictive because it
+> imposes a strong typing discipline, only works for full expressions, and in
+> its original version does not allow for code inspection. Lionel's Parreaux's
+> Squid system (paper to appear at upcoming POPL) is more powerful in that it
+> does allow inspection.
+
+Lionel Parreaux ([@LPTK]) is a PhD student at EPFL who has been working on
+related problems for the past couple years as part of his research on the
+intersection of database systems, programming languages and large-scala data
+analysis.
+
+## Squid: reusable and type-safe quasiquotes
+
+[Squid] is a metaprogramming framework that facilitates the type-safe
+manipulation of Scala programs through quasiquotes.
+Unlike scala-reflect quasiquotes, Squid quasiquotes are statically guaranteed
+to be well-typed, well-scoped and hygienic.
+The properties of Squid quasiquotes make them ideal for robust and safe
+transformation of trees, the exact problem we struggled with in our prototype
+macro system.
+
+One important attribute of Squid quasiquotes is that they are reusable.
+Similar to scala-reflect quasiquotes, it's possible to write generic utility
+functions with Squid quasiquotes and compose them into more complex tree
+structures.
+Many alternative statically typed quasiquote systems do not support this
+ability to combine small quasiquote fragments into larger programs.
+This functionality is enabled by Squid's first-class encoding of free
+variables.
+
+Squid quasiquotes statically track the context/scope of a quasiquote in its type to
+prevent references to unbound identifiers.
+For example, `code"(x: Int) + 1"` would be rejected at compile-time since `x`
+is undefined.
+Instead, quasiquotes must explicitly annotate free variables
+using question marks `code"(?x: Int) + 1)`.
+Free variables are then tracked in the second parameter of quasiquote's type
+`Code[Int, {x: Int}]`.
+
+To learn more about Squid works, I highly recommend reading their upcoming POPL
+2018 paper
+["Unifying Analytic and Statically-Typed Quasiquotes"](https://infoscience.epfl.ch/record/232427).
+Here is just one interesting table from the paper that I want to include here
+
+![Squid quasiquotes]({{ site.baseurl }}/resources/img/squid-quasiquotes.png)
+
+There are still many open questions on how to incorporate Squid with the new macros.
+For example,
+
+- Squid quasiquotes eagerly convert compiler trees to Squid's IR. This may
+  introduce problems for syntax like patterns and important metadata may get
+  lost like positions or attachments.
+- Squid currently does not support the ability to define methods or classes in
+  quasiquotes.
+
+I believe these problems are solvable, and we are discussing with
+Lionel and colleagues from and his lab about potential collaboration.
```

Comment:
`from and` -> `from`
`about potential collaboration` -> `about a potential collaboration` or `about potential collaborations`

Reply:
Fixed.